components/add_groups_to_channel_modal/add_groups_to_channel_modal.jsx

```diff
@@ -0,0 +1,278 @@
+import {FormattedMessage} from 'react-intl';
+
+import {Groups} from 'mattermost-redux/constants';
+
+import Constants from 'utils/constants.jsx';
+import {localizeMessage} from 'utils/utils.jsx';
+
+import MultiSelect from 'components/multiselect/multiselect.jsx';
+import groupsAvatar from 'images/groups-avatar.png';
+import AddIcon from 'components/icon/add_icon';
+
+const GROUPS_PER_PAGE = 50;
+const MAX_SELECTABLE_VALUES = 10;
+
+export default class AddGroupsToChannelModal extends React.Component {
+    static propTypes = {
+        currentChannelName: PropTypes.string.isRequired,
+        currentChannelId: PropTypes.string.isRequired,
+        searchTerm: PropTypes.string.isRequired,
+        groups: PropTypes.array.isRequired,
+        onHide: PropTypes.func,
+        actions: PropTypes.shape({
+            getGroupsNotAssociatedToChannel: PropTypes.func.isRequired,
+            setModalSearchTerm: PropTypes.func.isRequired,
+            linkGroupSyncable: PropTypes.func.isRequired,
+            getAllGroupsAssociatedToChannel: PropTypes.func.isRequired,
+        }).isRequired,
+    }
+
+    constructor(props) {
+        super(props);
+
+        this.searchTimeoutId = 0;
+
+        this.state = {
+            values: [],
+            show: true,
+            search: false,
+            saving: false,
+            addError: null,
+            loadingGroups: true,
+        };
+    }
+
+    componentDidMount() {
+        Promise.all([
+            this.props.actions.getGroupsNotAssociatedToChannel(this.props.currentChannelId, '', 0, GROUPS_PER_PAGE * 2),
+            this.props.actions.getAllGroupsAssociatedToChannel(this.props.currentChannelId),
+        ]).then(() => {
+            this.setGroupsLoadingState(false);
+        });
+    }
+
+    UNSAFE_componentWillReceiveProps(nextProps) { // eslint-disable-line camelcase
+        if (this.props.searchTerm !== nextProps.searchTerm) {
+            clearTimeout(this.searchTimeoutId);
+
+            const searchTerm = nextProps.searchTerm;
+            if (searchTerm === '') {
+                return;
+            }
+
+            this.searchTimeoutId = setTimeout(
+                async () => {
+                    this.setGroupsLoadingState(true);
+                    await this.props.actions.getGroupsNotAssociatedToChannel(this.props.currentChannelId, searchTerm);
+                    this.setGroupsLoadingState(false);
+                },
+                Constants.SEARCH_TIMEOUT_MILLISECONDS
+            );
+        }
+    }
+
+    handleHide = () => {
+        this.props.actions.setModalSearchTerm('');
+        this.setState({show: false});
+    }
+
+    handleExit = () => {
+        if (this.props.onHide) {
+            this.props.onHide();
+        }
+    }
+
+    handleResponse = (err) => {
+        let addError = null;
+        if (err && err.message) {
+            addError = err.message;
+        }
+
+        this.setState({
+            saving: false,
+            addError,
```

Comment:
nit: this could be simplified to:

```js
this.setState({
  saving: false,
  addError: (err && err.message) ? err.message : null,
});

Reply:
Is for things like this that I miss the [optional chaining proposal](https://github.com/tc39/proposal-optional-chaining):

```js
this.setState({
  saving: false,
  addError: err?.message
})
```